main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1071 @@
+
+      case Expr.NewChannel(exp1, exp2, tvar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val regionType = Type.mkRegion(regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = regionType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe2, exp2.loc)
+        val resTpe = tvar
+        val resEff = Type.mkUnion(eff1, eff2, regionVar, loc)
+        (resTpe, resEff)
+
+      case Expr.GetChannel(exp, tvar, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val channelType = Type.mkReceiver(elmVar, regionVar, loc)
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(expected = channelType, actual = tpe, exp.loc)
+        c.unifyTypeM(tvar, elmVar, loc)
+        val resTpe = tvar
+        val resEff = Type.mkUnion(eff, regionVar, loc)
+        (resTpe, resEff)
+
+      case Expr.PutChannel(exp1, exp2, loc) =>
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+        val elmVar = Type.freshVar(Kind.Star, loc)
+        val channelType = Type.mkSender(elmVar, regionVar, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = channelType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe2, exp2.loc)
+        val resTpe = Type.mkUnit(loc)
+        val resEff = Type.mkUnion(eff1, eff2, regionVar, loc)
+        (resTpe, resEff)
+
+      case Expr.SelectChannel(rules, default, tvar, loc) =>
+
+        val regionVar = Type.freshVar(Kind.Eff, loc)
+
+        /**
+          * Generates constraints for the SelectChannelRule.
+          */
+        def visitSelectRule(sr0: KindedAst.SelectChannelRule): (Type, Type) = {
+          sr0 match {
+            case KindedAst.SelectChannelRule(sym, chan, body) =>
+              val (chanType, eff1) = visitExp(chan)
+              val (bodyType, eff2) = visitExp(body)
+              c.unifyTypeM(chanType, Type.mkReceiver(sym.tvar, regionVar, sym.loc), sym.loc)
+              val resTpe = bodyType
+              val resEff = Type.mkUnion(eff1, eff2, regionVar, loc)
+              (resTpe, resEff)
+          }
+        }
+
+        /**
+          * Generates constraints for the default rule.
+          */
+        def visitDefaultRule(exp0: Option[KindedAst.Expr]): (Type, Type) =
+          exp0 match {
+            case None => (Type.freshVar(Kind.Star, loc), Type.Pure)
+            case Some(exp) => visitExp(exp)
+          }
+
+        val (ruleTypes, ruleEffs) = rules.map(visitSelectRule).unzip
+        val (defaultType, eff2) = visitDefaultRule(default)
+        c.unifyAllTypesM(tvar :: defaultType :: ruleTypes, Kind.Star, loc)
+        val resTpe = tvar
+        val resEff = Type.mkUnion(regionVar :: eff2 :: ruleEffs, loc)
+        (resTpe, resEff)
+
+      case Expr.Spawn(exp1, exp2, loc) =>
```

Comment:
We need a comment here that we don't fully understand yet what the type of spawn should be.